ARCs/arc-0004.md

```diff
@@ -0,0 +1,498 @@
+opt-in would require.  If an opt-in or close-out contains arguments,
+an ARC-4 app **MUST** fail the transaction, or interpret the arguments
+as an invocation as described in the rest of this document, to perform
+along with the opt-in or close-out.
+
+### Foreign apps, asas, and accounts
+
+Some AVM opcodes require specific values to be placed in the "foreign
+arrays" of the application call transaction. Three special types,
+`account`, `asset`, `application` allow methods to describe these
+requirements. To encode method calls with such types, place the value
+in question in the Accounts (`apat`), Foreign Assets (`apas`), or
+Foreign Assets (`apfa`) arrays, and encode a uint8 referring to that
+value in the normal location for the value. For the Accounts and
+Applications array, a 0 implicitly refers to the Sender and the called
+application respectively, and the passed values begin at 1.
+
+When passing addresses, ASAs, or apps that are _not_ required to be
+accessed by such opcodes, ARC-4 Apps **SHOULD** use the base types for
+passing these types: `address` and `uint64`.
+
+### Transaction groups.
+
+Some apps require that they are invoked as part of a larger
+transaction group, containing specific additional transactions. Future
+versions of this document may describe how a method signature could
+describe these requirements.
+
+ARC-4 Apps **SHOULD** be constructed to allow their invocations to be
+combined in a single atomic group if they can do so safely. For
+example, they should use `txns` to examine the previous index for a
+required `pay` transaction, rather than hardcode an index with `txn`.
+
+> Comments on this section are particularly welcome. A strawman
+> proposal is to introduce a new type for each transaction type. The
+> appearance of such a transaction type would imply that such a
+> transaction must appear in the transaction group, in the order of
+> their appearance in the call. Unlike "foreign" references, these
+> special types would not be encoded as a small integer "pointing" to
+> the associated object, as allowing such references would create
+> opportunities for multiple application calls to point to the same
+> transaction in the group, which is undesirable. Instead, the
+> locations of the transactions would be implied entirely by the
+> placement in the argument list.
+
+> Further questions: Should it be possible to describe requirements of
+> certain fields of the transactions named this way? For example, can
+> a signature specify the setting of the the size or recipient of a pay
+> transaction? A specific, important instance of this problem: How can
+> logicsigs on these transactions be provided in the JSON interface?
+
+## New Features in support of the ABI
+
+Some changes are needed to the Algorand protocol and its tools to
+better support this ABI:
+
+* The `log` opcode can be used to return a value. A convention for
+  disguishing the return value from other logs is required.
```

Comment:
We've add a bit saying that return values are logged as concat(hash("return")||encode(value)). So the prefix would always be hash("return") = 0x151f7c75.

For events, I agree we'd want a name involved, but I don't see the need for "return".  What do you think?

(This would also allow us to give you an opcode `approve` that logs the top-of-stack with the return prefix, and returns success.)